prosac_from_c.py

```diff
@@ -0,0 +1,143 @@
+def prosac(sorted_matches):
+    # Adapted from: http://devernay.free.fr/vision/src/prosac.c	
+    # TODO: move this distCoeffs out!
+    distCoeffs = np.zeros((5, 1))  # assume zero for now
+
+    CORRESPONDENCES = sorted_matches.shape[0]
+    # isInlier = np.zeros([1,CORRESPONDENCES])
+    SAMPLE_SIZE = 4
+    MAX_OUTLIERS_PROPORTION = 0.8
+    P_GOOD_SAMPLE = 0.99
+    TEST_NB_OF_DRAWS = 60000
+    TEST_INLIERS_RATIO = 0.5
+    BETA = 0.1
+    ETA0 = 0.05
+
+    def niter_RANSAC(p, epsilon, s, Nmax):
+        if(Nmax == -1):
+            Nmax = np.iinfo(np.int32).max
+        if(not (Nmax >= 1)):
+            print("C++ Assertion failed - 1")
+        if(epsilon <= 0):
+            return 1
+        logarg = - np.exp(s * np.log(1 - epsilon))
+        logval = np.log(1 + logarg)
+        N = np.log(1 - p) / logval
+        if(logval < 0 and N < Nmax):
+            return np.ceil(N)
+
+        return Nmax
+
+    def Imin(m, n, beta):
+        mu = n*beta
+        sigma = np.sqrt(n * beta * (1 - beta))
+        return  np.ceil(m + mu + sigma * np.sqrt(2.706))
+
+    N = CORRESPONDENCES
+    m = SAMPLE_SIZE
+    T_N = niter_RANSAC(P_GOOD_SAMPLE,MAX_OUTLIERS_PROPORTION,SAMPLE_SIZE,-1)
+    beta = BETA
+    I_N_min = (1 - MAX_OUTLIERS_PROPORTION)*N
+    logeta0 = np.log(ETA0)
+
+    # print("PROSAC sampling test\n")
+    # print("number of correspondences (N):{0}\n".format(N))
+    # print("sample size (m):{0}\n".format(m))
+    # print("showing the first {0} draws from PROSAC\n".format(TEST_NB_OF_DRAWS))
+
+    n_star = N
+    I_n_star = 0
+    I_N_best = 0
+    t = 0
+    n = m
+    T_n = T_N
+
+    for i in range(m):
+        T_n = T_n * (n - i) / (N - i)
+
+    T_n_prime = 1
+    k_n_star = T_N
+
+    best_model = {}
+    while (((I_N_best < I_N_min) or t <= k_n_star) and t < T_N and t <= TEST_NB_OF_DRAWS):
+        model = {}
+        inliers = []
+        t = t + 1
+        # print("Iteration t=%d, " % t)
+
+        if ((t > T_n_prime) and (n < n_star)):
+            T_nplus1 = (T_n * (n+1)) / (n+1-m)
+            n = n + 1
+            T_n_prime = T_n_prime + np.ceil(T_nplus1 - T_n)
+            T_n = T_nplus1
+
+        if (t > T_n_prime):
+            pts_idx = np.random.choice(n, m, replace=False)
+        else:
+            pts_idx = np.append(np.random.choice(n-1, m-1, replace=False), n-1)
```

Comment:
@willGuimont Also here I think it should be` n-1` for the new point as it is an index for a 0-based array. If I just use `n` I get an out-of-bound index error..